lib/datadog/tracing/contrib/grpc/distributed/propagation.rb

```diff
@@ -0,0 +1,38 @@
+# frozen_string_literal: true
+# typed: true
+
+require_relative 'fetcher'
+require_relative '../../../distributed/b3'
+require_relative '../../../distributed/b3_single'
+require_relative '../../../distributed/datadog'
+require_relative '../../../distributed/propagation'
+
+module Datadog
+  module Tracing
+    module Contrib
+      module GRPC
+        module Distributed
+          # Extracts and injects propagation through gRPC metadata.
+          # @see https://github.com/grpc/grpc-go/blob/v1.50.1/Documentation/grpc-metadata.md
+          class Propagation < Tracing::Distributed::Propagation
+            def initialize
+              super(
+                propagation_styles: {
+                  Tracing::Configuration::Ext::Distributed::PROPAGATION_STYLE_B3 =>
+                    Tracing::Distributed::B3.new(fetcher: Fetcher),
+                  Tracing::Configuration::Ext::Distributed::PROPAGATION_STYLE_B3_SINGLE_HEADER =>
+                    Tracing::Distributed::B3Single.new(fetcher: Fetcher),
+                  Tracing::Configuration::Ext::Distributed::PROPAGATION_STYLE_DATADOG =>
+                    Tracing::Distributed::Datadog.new(fetcher: Fetcher)
+                })
+            end
+
+            # DEV: Singleton kept until a larger refactor is performed.
+            # DEV: See {Datadog::Tracing::Distributed::Propagation#initialize} for more information.
+            INSTANCE = Propagation.new
```

Comment:
Minor: I slightly wonder if we should take the next step and also kill the `GRPC::Distributed::Propagation` subclass, since it's only a shortcut for calling `Tracing::Distributed::Propagation` with the defaults :)

Reply:
I think this is complex enough that it should be encapsulated:
```ruby
                propagation_styles: {
                  Tracing::Configuration::Ext::Distributed::PROPAGATION_STYLE_B3 =>
                    Tracing::Distributed::B3.new(fetcher: Fetcher),
                  Tracing::Configuration::Ext::Distributed::PROPAGATION_STYLE_B3_SINGLE_HEADER =>
                    Tracing::Distributed::B3Single.new(fetcher: Fetcher),
                  Tracing::Configuration::Ext::Distributed::PROPAGATION_STYLE_DATADOG =>
                    Tracing::Distributed::Datadog.new(fetcher: Fetcher)
                })
```
It's the one place where we wire the protocol-specific logic.
I'd like to request permission to leave it as it.